aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,953 @@
+    """
+    Input: '()'
+    Output: []
+    """
+    if return_decl == '()':
+        return []
+    if return_decl[0] == '(' and return_decl[-1] == ')':
+        return_decl = return_decl[1:-1]
+    returns = []
+    for arg in return_decl.split(', '):
+        returns.append(Argument.parse(arg))
+    return returns
+
+def parse_arguments(args: str) -> Tuple[Sequence[Argument], Sequence[Argument], Sequence[Argument]]:
+    """
+    Input: 'int x, int y, int z'
+    Output: positional args, kwarg only args
+    """
+    arguments: List[Argument] = []
+    kwarg_only_arguments: List[Argument] = []
+    out_arguments: List[Argument] = []
+    arguments_acc = arguments
+
+    # TODO: Use a real parser here; this will get bamboozled
+    # by signatures that contain things like std::array<bool, 2> (note the space)
+    for arg in args.split(', '):
+        if not arg:
+            continue
+        if arg == '*':
+            assert arguments_acc is arguments, "invalid syntax: kwarg-only specifier * can only occur once"
+            arguments_acc = kwarg_only_arguments
+            continue
+        parg = Argument.parse(arg)
+        # Currently, we rely directly on the invariant that there are NO
+        # kwarg-only mutating arguments.  If you want to relax this,
+        # we will need a more semantic way of matching that takes
+        # into account return arguments.  In that case, you will have
+        # to manage out_arguments computation a level up, in
+        # FunctionSchema.  See Note [is_out_fn]
+        if parg.annotation is not None and parg.annotation.is_write:
+            if arguments_acc is arguments:
+                pass  # do nothing
+            elif arguments_acc is kwarg_only_arguments:
+                arguments_acc = out_arguments
+        else:
+            assert arguments_acc is not out_arguments
+        arguments_acc.append(parg)
+
+    return arguments, kwarg_only_arguments, out_arguments
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                           PROCESSING
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# Conveniently add error context to exceptions raised.  Lets us
+# easily say that an error occurred while processing a specific
+# context.
+@contextlib.contextmanager
+def context(msg: str) -> Iterator[None]:
+    try:
+        yield
+    except Exception as e:
+        # TODO: this does the wrong thing with KeyErorr
+        msg = textwrap.indent(msg, '  ')
+        msg = f'{e.args[0]}\n{msg}' if e.args else msg
+        e.args = (msg,) + e.args[1:]
+        raise
+
+# Represent a source location; used for better error reporting
+@dataclass
+class Location:
+    file: str
+    line: int
+
+    def __str__(self) -> str:
+        return "{}:{}".format(self.file, self.line)
+
+# A custom loader for YAML to let us also keep track of line numbers
+# of each entry in the YAML file
+class LineLoader(Loader):
```

Comment:
This is gold. I haven't looked, but if you can save location info into Declarations.yaml it could be used in the tools/ codegen for the same usability bump (although at least some of that code is pretty careful to hand-introduce some context - op names mostly IIRC - into its errors, so it's not awful. But this would be better).